elide-datastore/elide-datastore-aggregation/src/main/java/com/yahoo/elide/datastores/aggregation/queryengines/sql/SQLQueryConstructor.java

```diff
@@ -0,0 +1,507 @@
+     *
+     * @param groupByDimensions columns to project out
+     * @param queriedTable queried analytic view
+     * @return <code>GROUP BY tb1.col1, tb2.col2, ...</code>
+     */
+    private String referenceProjectGroupBy(Set<ColumnProjection> groupByDimensions, SQLAnalyticView queriedTable) {
+        return "GROUP BY " + groupByDimensions.stream()
+                .map(dimension -> resolveSQLColumnReference(dimension, queriedTable))
+                .collect(Collectors.joining(", "));
+    }
+
+    /**
+     * Construct HAVING clause filter using physical column references. Metric fields need to be aggregated in HAVING.
+     *
+     * @param predicate a filter predicate in HAVING clause
+     * @param table Elide logical table this query is querying
+     * @param template query template
+     * @return an filter/constraint expression that can be put in HAVING clause
+     */
+    private String referenceProjectHavingFilter(FilterPredicate predicate, Table table, SQLQueryTemplate template) {
+        Path.PathElement last = predicate.getPath().lastElement().get();
+        Class<?> lastClass = last.getType();
+        String fieldName = last.getFieldName();
+
+        if (!lastClass.equals(table.getCls())) {
+            throw new InvalidPredicateException("The having clause can only reference fact table aggregations.");
+        }
+
+        MetricFunctionInvocation metric = template.getMetrics().stream()
+                // TODO: filter predicate should support alias
+                .filter(invocation -> invocation.getAlias().equals(fieldName))
+                .findFirst()
+                .orElse(null);
+
+        if (metric != null) {
+            return metric.getFunctionExpression();
+        } else {
+            return generateColumnReference(predicate);
+        }
+    }
+
+    /**
+     * Construct SELECT statement expression with metrics and dimensions directly using physical table column
+     * references.
+     *
+     * @param template query template with nested subquery
+     * @param queriedTable queried analytic view
+     * @return <code>SELECT function(metric1) AS alias1, tb1.dimension1 AS alias2</code>
+     */
+    private String referenceProject(SQLQueryTemplate template, SQLAnalyticView queriedTable) {
+        // TODO: project metric field using table column reference
+        List<String> metricProjections = template.getMetrics().stream()
+                .map(invocation -> invocation.getFunctionExpression() + " AS " + invocation.getAlias())
+                .collect(Collectors.toList());
+
+        List<String> dimensionProjections = template.getGroupByDimensions().stream()
+                .map(dimension ->  resolveSQLColumnReference(dimension, queriedTable) + " AS " + dimension.getAlias())
+                .collect(Collectors.toList());
+
+        String projectionClause = String.join(",", metricProjections);
+
+        if (!dimensionProjections.isEmpty()) {
```

Comment:
You can just stream concat the two lists and then use the join operator with comma.

Reply:
👍